runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java

```diff
@@ -0,0 +1,381 @@
+		bankActive = false;
+		log.debug("Bank value stopping");
+
+		if (executor != null)
+		{
+			executor.shutdownNow();
+		}
+
+		if (bankThread != null)
+		{
+			bankThread.interrupt();
+		}
+	}
+
+	private void calculateBankWithThread()
+	{
+		bankActive = true;
+		log.debug("Bank value running");
+		bankThread = new Thread(this::calculateBank);
+		bankThread.start();
+	}
+
+	class ItemPriceAndAlch extends ItemPrice
+	{
+		@Getter
+		@Setter
+		private int highAlchPrice;
+
+		ItemPriceAndAlch(int gePrice, int highAlchPrice)
+		{
+			setPrice(gePrice);
+			this.highAlchPrice = highAlchPrice;
+		}
+	}
+
+	private final Object rsItemLock = new Object();
+
+	/**
+	 * Grabs the GE price, calculates the high alch price, then multiplies the quantity
+	 * @param item
+	 * @return
+	 */
+	private ItemPriceAndAlch getItemPrices(WidgetItem item)
+	{
+		if (COINS == item.getId())
+		{
+			return new ItemPriceAndAlch(item.getQuantity(), item.getQuantity());
+		}
+
+		ItemComposition itemComposition;
+
+		// Having this lock seems to not crash runescape. I'm not sure if it's because of this method call
+		// or if it's something else that this lock is slowing down. I believe it's because this is the only
+		// place that has a possibility of accessing runescape's code.
+		synchronized (rsItemLock)
+		{
+			itemComposition = itemManager.getItemComposition(item.getId());
+		}
+
+		if (itemComposition == null)
+		{
+			return new ItemPriceAndAlch(0, 0);
+		}
+
+		ItemPrice itemPrice = null;
+
+		if (config.showGE() && itemComposition.isTradable())
+		{
+			try
+			{
+				itemPrice = itemManager.getItemPrice(itemComposition.getId());
+			}
+			catch (IOException e)
+			{
+				return new ItemPriceAndAlch(0, 0);
+			}
+		}
+
+		int gePrice = 0;
+
+		if (config.showGE() && itemPrice != null)
+		{
+			gePrice = itemPrice.getPrice() * item.getQuantity();
+		}
+
+		int itemCompositionPrice = itemComposition.getPrice();
+
+		int highAlchPrice = 0;
+
+		if (itemCompositionPrice >= 0)
+		{
+			highAlchPrice = Math.round(itemCompositionPrice * HIGH_ALCHEMY_CONSTANT) * item.getQuantity();
+		}
+
+		return new ItemPriceAndAlch(gePrice, highAlchPrice);
+	}
+
+	private void calculateBank()
+	{
+		AtomicLong gePrice = new AtomicLong(0);
+		AtomicLong haPrice = new AtomicLong(0);
+
+		Widget widgetBankTitleBar = client.getWidget(WidgetInfo.BANK_TITLE_BAR);
+
+		if (widgetBankTitleBar == null || widgetBankTitleBar.isHidden())
+		{
+			return;
+		}
+
+		// Another hack: wait 1 second for the GUI to come up (to properly update the bank title and grab loaded items)
+		// If we can find the correct event, we may be able to remove this
+		try
+		{
+			Thread.sleep(1000);
+		}
+		catch (InterruptedException e)
+		{
+			// happens when spamming config settings and opening and closing bank quickly
+			return;
+		}
+
+		WidgetItem[] widgetItems = new BankItemQuery().result(client);
+
+		if (widgetItems.length == 0)
+		{
+			return;
+		}
+
+		executor = Executors.newCachedThreadPool();
```

Comment:
I would probably make an api method on the web service that took in a list of itemids and mass-looked their prices up at once, so it could be in one request. And then maybe bypass item manager completely and use ItemClient.

Reply:
That would be definitely a lot better than creating new executor in separate thread and submitting bunch of tasks into it, that can't be correct no matter how I look at it.